include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -344,9 +694,484 @@ class alignment_file_format_sam
     //!\brief The format version string.
     static constexpr char format_version[4] = "1.6";
 
-    //!\brief A variable that tracks whether the header_ptr as been written or not.
+    //!\brief A buffer used when parsing arithmetic values with std::from_chars.
+    std::array<char, 316> buffer{}; // Doubles can be up to 316 characters
+
+    //!\brief A variable that tracks whether the content of header has been written or not.
     bool written_header{false};
 
+    //!\brief Tracks whether reference information (\@SR tag) were found in the header
+    bool ref_info_present_in_header{false};
+
+    /*!\brief Checks for known reference ids or adds a new reference is and assigns a reference id to `ref_id`.
+     * \tparam ref_id_type             The type of the reference id (usually a view::all over ref_id_tmp_type).
+     * \tparam ref_id_tmp_type         The type of the temporary parsed id (Same type as reference ids in header).
+     * \tparam header_type             The type of the alignment header.
+     * \tparam ref_seqs_type  A tag whether the reference information were given or not (std::ignore or not).
+     *
+     * \param[in, out] ref_id      The reference id to be filled.
+     * \param[in, out] ref_id_tmp  The temporary of the parsed reference id.
+     * \param[in, out] header      The header object that stores the reference id information.
+     */
+    template <typename ref_id_type,
+              typename ref_id_tmp_type,
+              typename header_type,
+              typename ref_seqs_type>
+    void check_and_assign_ref_id(ref_id_type            & ref_id,
+                                 ref_id_tmp_type        & ref_id_tmp,
+                                 header_type            & header,
+                                 ref_seqs_type & /*tag*/)
+    {
+        if (!std::ranges::empty(ref_id_tmp)) // otherwise the std::optional will not be filled
+        {
+            auto search = header.ref_dict.find(ref_id_tmp);
+
+            if (search == header.ref_dict.end())
+            {
+                if constexpr(detail::decays_to_ignore_v<ref_seqs_type>) // no reference information given
+                {
+                    if (ref_info_present_in_header)
+                    {
+                        throw format_error{"Unknown reference id found in record which is not present in the header."};
```

Comment:
```suggestion
                        throw format_error{std::string{"Encountered reference ID \""} + ref_id_tmp + "\" in a record, but this was not present in the header."};
```

Reply:
technically the type could be something else than a string... ince we may type deduct the ref_ids type

Reply:
Hm, let's make a card to re-evaluate the usefulness of exception in I/O at a alter point in time.

Reply:
I added a card